### PR TITLE
Function Pointer ObjectBehaviorDefinition Hooks

### PR DESCRIPTION
**Pull request by ShellyHerself**
Basically handles #39 

To actually do any hooks with this now there just needs to be a function created that has the same attributes as the function being hooked. And you just assign the pointer to that function to the struct entry.

If you actually want to execute the original code with that you should also call the original function, which you do by calling \<type\>_<function_name>

### CODE REVIEW

File: halo/hooks/object.cpp [hunk 1]
```diff
@@ -28,9 +32,240 @@ Signature(true, sig_biped_jump,
...
+    new_ssce_beh.parent_definitions[1] = &new_ssce_beh;
+
+    // Backup old def pointers.
+
+    memcpy(&backup, game_defs, sizeof(ObjectBehaviorDefinition));
+
+    // Replace old defs with our new defs.
+
+    DWORD prota, protb;
+    VirtualProtect(game_defs, 4*12, PAGE_EXECUTE_READWRITE, &prota);
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: What is this magic number `4*12`?

File: halo/hooks/object.cpp [hunk 2]
```diff
@@ -28,9 +32,240 @@ Signature(true, sig_biped_jump,
...
+    // End of Macro
+
+static ObjectBehaviorDefinition* backup[12];
 
 void init_object_hooks(){
+    auto game_defs = object_behavior_defs();
+
+    // Copy data to our new defs.
+
+    memcpy(&new_obje_beh, game_defs[static_cast<int>(ObjectType::BIPED)]->parent_definitions[0],  sizeof(ObjectBehaviorDefinition));
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: Do we really need to `static_cast<int>` the enum to get its integer value?

Comment (Mimickal) [inline halo/hooks/object.cpp]: You could shorten all of these lines by storing the value of `sizeof(ObjectBehaviorDefinition)` in a variable

File: halo/hooks/object.cpp [hunk 3]
```diff
@@ -28,9 +32,240 @@ Signature(true, sig_biped_jump,
     {0x8B, 0x0D, -1, -1, -1, 0x00,
      0x8B, 0x51, 0x34, 0x83, 0xEC, 0x10, 0x53, 0x55}); // size 6 do manual
 
+#define OBJECT_BEHAVIOR_FUNCTION_PTRS(type) \
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: Minor stylistic nitpick: `OBJECT_BEHAVIOR_FUNCTION_PTRS` is a noun, but this name is being used for a function, so it should be a verb. What does this function actually do? Would something like `DECLARE_OBJECT_BEHAVIOR_FUNCTION_PTRS` make more sense?

File: halo/hooks/object.cpp [hunk 4]
```diff
@@ -5,9 +5,13 @@
...
 #include "../../hooker/hooker.hpp"
-#include "../memory/object.hpp"
+#include "../../hooker/function_pointer_safe.hpp"
+#include "../memory/gamestate/object/object.hpp"
+#include "../memory/behavior_definition.hpp"
+#include <windows.h>
+#include <cstring>
 
-Signture(true, sig_object_update,
+Signature(true, sig_object_update,
```
Comment (Mimickal) [inline halo/hooks/object.cpp:18]: Why are these values significant? A comment explaining these values would be very helpful here.

File: halo/hooks/object.cpp [hunk 5]
```diff
@@ -28,9 +32,240 @@ Signature(true, sig_biped_jump,
...
+    ObjRenderDebug                type ## _render_debug                  = NULL; \
+    ObjNewToNetwork               type ## _new_to_network                = NULL; \
+    ObjUpdateBaseline             type ## _update_baseline               = NULL; \
+    ObjGenerateUpdateHeader       type ## _generate_update_header        = NULL; \
+    ObjHandleReceivedUpdate       type ## _handle_received_update        = NULL; \
+    ObjShouldDoUpdateIncremental  type ## _should_do_update_incremental  = NULL; \
+    ObjSetLastUpdateTime          type ## _set_last_update_time          = NULL; \
+    // End of Macro
+
+ObjectBehaviorDefinition new_obje_beh;
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: From our voice conversation --These values are all enumerated out (instead of being an array, for example) for easier access via assembly? If so, that's an important note to put here.

**Review by Mimickal — CHANGES_REQUESTED** [review]

File: halo/hooks/object.cpp (hunk 4 above)
Comment (ShellyHerself) [reply halo/hooks/object.cpp:18]: I'll instead rename them to be more clear in the next pull request where they are actually used. Then I'll document what a signature is and what it's for in a readme down the line.

Comment (ShellyHerself) [reply halo/hooks/object.cpp:18]: I'll also update the comments there so they actually mean something.

File: halo/hooks/object.cpp (hunk 1 above)
Comment (ShellyHerself) [reply halo/hooks/object.cpp]: Addressed by ce0ec43

File: halo/hooks/object.cpp [hunk 6]
```diff
@@ -146,116 +147,119 @@ extern "C" { // Demangle naming for ASM.
...
+static ObjectBehaviorDefinition* vanilla_def_pointers_backup[POSITIVE_OBJECT_TYPES];
+const size_t OBJECT_DEF_PTR_ARRAY_ABS_SIZE = sizeof(uintptr_t)*POSITIVE_OBJECT_TYPES;
 
 void init_object_hooks(){
     auto game_defs = object_behavior_defs();
 
     // Copy data to our new defs.
 
-    const size_t s = sizeof(ObjectBehaviorDefinition);
+    const size_t S = sizeof(ObjectBehaviorDefinition);
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: Stylistic nitpick:
1. Single character names are very hard to parse. Even a name like `obd_size` would be much better here.
2. While this value is technically a constant, you're defining it inline partway through a function, so you don't need to use CAPITAL_SNAKE_CASE, or `const`.

File: halo/hooks/object.cpp [hunk 7]
```diff
@@ -146,116 +147,119 @@ extern "C" { // Demangle naming for ASM.
...
 
     // These top 4 are negative indices and are not in the base array,
     // but we still need them. They are parent types of other object types
     // found here though, so we get that data by looking through the parent
     // type definitions.
 
-    memcpy(&new_obje_beh, game_defs[static_cast<int>(ObjectType::BIPED)]->parent_definitions[0], s);
+    memcpy(&new_obje_beh, game_defs[etoi(ObjectType::BIPED)]->parent_definitions[0], S);
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: This is much nicer :+1:

File: halo/hooks/object.cpp [hunk 8]
```diff
@@ -146,116 +147,119 @@ extern "C" { // Demangle naming for ASM.
...
-    memcpy(&new_ssce_beh, game_defs[static_cast<int>(ObjectType::SOUND_SCENERY)], s);
+    memcpy(&new_ssce_beh, game_defs[etoi(ObjectType::SOUND_SCENERY)], S);
     new_ssce_beh.parent_definitions[0] = &new_obje_beh;
     new_ssce_beh.parent_definitions[1] = &new_ssce_beh;
 
-    // Backup old def pointers for when the DLL needs to unload.
+    // Backup old array of def pointers for when the DLL needs to unload.
 
-    memcpy(&backup, game_defs, sizeof(ObjectBehaviorDefinition));
+    const size_t COPY_SIZE = OBJECT_DEF_PTR_ARRAY_ABS_SIZE;
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: Why redefine this here when the constant at the top of the file exists? I'd argue `OBJECT_DEF_PTR_ARRAY_ABS_SIZE` makes the intent more clear here than copy size, especially considering the other names involved in the below `memcpy` call.

File: halo/hooks/object.cpp [hunk 9]
```diff
@@ -293,9 +297,11 @@ void revert_object_hooks(){
 
     // Revert to the old defs.
 
+    const size_t COPY_SIZE = OBJECT_DEF_PTR_ARRAY_ABS_SIZE;
```
Comment (Mimickal) [inline halo/hooks/object.cpp]: Same here

**Review by Mimickal — CHANGES_REQUESTED** [review]

**Review by Mimickal — APPROVED** [review]
This is infinitely more understandable with the added comments. Nicely done :+1:
